lib/rouge/lexers/cython.rb

```diff
@@ -0,0 +1,68 @@
+# -*- coding: utf-8 -*- #
+# frozen_string_literal: true
+
+module Rouge
+  module Lexers
+    load_lexer 'python.rb'
+
+    class Cython < Python
+      title "Cython"
+      desc "Cython and Pyrex source code (cython.org)"
+      tag 'cython'
+      aliases 'pyx', 'pyrex'
+      filenames '*.pyx', '*.pxd', '*.pxi'
+      mimetypes 'text/x-cython', 'application/x-cython'
+
+      def self.keywords
+        @keywords ||= super + Array.new(%w(
+          by ctypedef except? fused gil nogil 
+        ))
+      end
+
+      dotted_identifier = /[a-z_.][a-z0-9_.]*/i
+      prepend :root do
+        rule %r/(cimport)(\s+)(#{dotted_identifier})/ do
+          groups Keyword::Namespace, Text, Name::Namespace
+        end
+
+        rule %r/(from)((?:\\\s|\s)+)(#{dotted_identifier})((?:\\\s|\s)+)(cimport)/ do
+          groups Keyword::Namespace,
+                 Text,
+                 Name::Namespace,
+                 Text,
+                 Keyword::Namespace
+        end
+
+        rule %r/(cp?def)(\s+)/ do
+          groups Keyword, Text
+          push :cdef
+        end
+
+        rule %r/(cdef)(:)/ do
+          groups Keyword, Punctuation
+        end
+
+        rule %r/(struct)((?:\s|\\\s)+)/ do
+          groups Keyword, Text
+          push :classname
+        end
+      end
+
+      state :cdef do
+        rule %r/(public|readonly|extern|api|inline)\b/, Keyword::Reserved
+        rule %r/(struct|enum|union|class)\b/, Keyword
+        rule %r/([a-zA-Z_]\w*)(\s*)(?=[(:#=]|$)/ do 
+          groups Name::Function, Text
+          :pop!
+        end
+        rule %r/([a-zA-Z_]\w*)(\s*)(,)/ do
+          groups Name::Function, Text, Punctuation
+        end
+        rule %r/from\b/, Keyword, :pop!
+        rule %r/as\b/, Keyword
+        mixin :root
```

Comment:
Do you really want to mix in the rules from `:root`? Isn't all you really need just a whitespace rule?

Reply:
So I was a little unsure how to handle this (this is where my lack of language knowledge hurts a bit).

The parent Python lexer doesn't define a specific whitespace class. When copying the pygments file, they manually add a selection of what looks like whitespace rules, which I initially replaced with the mixin.

```
(r':', Punctuation, '#pop'),
(r'(?=["\'])', Text, '#pop'),
(r'[a-zA-Z_]\w*', Keyword.Type),
(r'.', Text),
```

Using just the translated version of these rules caused issues with newline tokens in testing. I copied the newline rule from root to match this, leaving these rules.

```
        rule %r/:/, Punctuation, :pop!
        rule %r/(?=["\'])/, Text, :pop!
        rule %r/[a-zA-Z_]\w*/, Keyword::Type
        rule %r/\n+/m, Text
        rule %r/./, Text
```

I actually think I was missing a case here too by omitting the first 2 rules that pop on ':' and assignment.

I am a little confused as to why the last rule doesn't match newlines - is there a cleaner way to write these out?

Alternatively I could add a whitespace definition to the python lexer, but I had been avoiding making changes to that file.

Reply:
I fixed this by adding the option to have '.' match newlines and removed the extra rule in the latest commit. The code should be good to go.

Reply:
@mwaddoups Thanks for all the updates so far! I'm sorry I didn't get a chance to save you the trouble but you got it: regular expression patterns in Ruby don't match newlines unless the `m` modifier is set on the expression (or unless there's an express ‘\n` character).

That said, while the code is no longer generating error tokens (which is good), if you look at the visual sample, you can see that there's a host of issues with the highlighting: types inside cdef blocks, punctuation like `[` and `]` and keywords like `for`. Some of these issues look to me like they come from the fact we're applying rules from the Pygments' Cython lexer to extend the Python lexer while it seems to redefine the entire lexer but some of them are present there, too (see [this this demo](http://pygments.org/demo/6856569/) where things like types are not highlighting consistently). 

As best I can tell, the cause of the problems is how to handle (what I'm calling) `cdef` blocks. The lexer should be treated these blocks as significant and moving into and out of states based on indentation but isn't. Instead, the rules from Pygments are essentially heuristics that, while being correct some of the time, have holes.

How do we deal with this? Well, Rouge has the capability to specify the indentation using the [Indentation](https://github.com/rouge-ruby/rouge/blob/7cdb1f9da96660afb144568c8b9029e474797ec9/lib/rouge/util.rb#L68) module. You can see examples of it in action in the [Sass](https://github.com/rouge-ruby/rouge/blob/f31a5eeba00c2f7ed9ee9eb4cc3f06bf9161df53/lib/rouge/lexers/sass.rb), [Haml](https://github.com/rouge-ruby/rouge/blob/768b10ece40b14043d56926e026ae8a07aec6120/lib/rouge/lexers/haml.rb) and [Slim](https://github.com/rouge-ruby/rouge/blob/f31a5eeba00c2f7ed9ee9eb4cc3f06bf9161df53/lib/rouge/lexers/slim.rb) lexers. Truthfully, I'm not that familiar with it but I suspect it might be the way to do this properly.

I'm a bit conflicted about whether we shouldn't just include _a_ Cython lexer and worry about perfecting it later but if you don't mind, I'd like to see how difficult it is to try this approach. If you want to have a crack at it, feel free, but I'll try to have a go myself sometime this week.

Reply:
Ah yes, good point that the highlighting is failing for complex cdef blocks.

Maybe this is naive, but I had a look and just adding a pop on newline to the cdef definition gets us 90% of the way there - this seems like the behaviour we want anyway. I added a more complex nested cdef as well. But this may create more problems I'm not aware of - certainly your suggested approach looks more elegant.

We still fail on highlighting post assignment operator - the line

`cdef int total = 0`

But I think with a cleverer regex this case can be popped off properly too (post equals sign).

Will be interested to see if you can make the more elegant approach work, it's beyond my skills for sure!

Reply:
Sorry—took a couple more days than anticipated!

I've just committed a patch that looks to me (at least per the visual sample) to be highlighting everything correctly. Could you have a look and see if it's missing something?

Reply:
Hmmm, I can see quite a few rough edges now that I'm looking at it here. Perhaps let me clean those up and add some comments to explain what I'm doing.

Reply:
OK, what's now been pushed is more robust. It handles the expanded visual sample fine but I'm keen to hear how it behaves when used with some more real world code. Please have a test and let me know if you spot any oversights.

Oh, and I've added some more comments to the lexer to explain what's happening. If anything is unclear, let me know!

Reply:
This looks really good to me - and a much more complete Cython lexer to boot as well! I've had a look at a few files from our cython codebase and the results look perfect! It's quite an expanded language so as always it's hard to guarantee full coverage but I would say it looks good to go.

Thanks for doing so much of the legwork on this - felt like all I had to really do was get the ball rolling!

Reply:
No worries. If you see any problems, you can always leave a bug report in Issues!